reports-backend/src/main/java/org/jboss/da/reports/impl/ReportsGeneratorImpl.java

```diff
@@ -81,35 +88,198 @@
+        return currentBestVersion;
+    }
+
+    private List<String> getAvailableWhitelistVersions(Map<WhiteArtifact, Set<ProductVersion>> allArtifacts,
+            GAV currentGav) {
+        Set<WhiteArtifact> whiteArtifactsForCurrentGav = allArtifacts.keySet().stream()
+                .filter(x -> {
+                    return x.getGa().getArtifactId()
+                            .equals(
+                           currentGav.getGA().getArtifactId())
+                            
+                            &&
+                           
+                           x.getGa().getGroupId()
+                            .equals(
+                           currentGav.getGA().getGroupId());
+                })
+                .collect(Collectors.toSet());
+                
+        List<String> whiteVersionsForCurrentGav = whiteArtifactsForCurrentGav.stream()
+                .map(x -> x.getVersion())
+                .collect(Collectors.toList());
+        
+        List<String> availableVersionsFromWhitelist = versionFinder.getBuiltVersionsFor(currentGav, whiteVersionsForCurrentGav);
+        return availableVersionsFromWhitelist;
+    }
+
+    private Set<Long> getProductVersionIds(Set<ProductVersion> relevantProductVersions) {
+        return relevantProductVersions.stream()
+                .map(x -> x.getId())
+                .collect(Collectors.toSet());
+    }
+
+    private Set<ProductVersion> getProductVersions(Set<String> productNames, Set<Long> productVersionIds){
+        Set<ProductVersion> productVersions = new HashSet<>();
+        
+        if(productNames != null){
+            for(String productName : productNames){
+                List<ProductVersion> prodVers = productVersionService.getAllForProduct(productName); 
+                productVersions.addAll(prodVers);
+            }
+        }
+        
+        if(productVersionIds != null){
+            for(Long prodVerId : productVersionIds){
+                Optional<ProductVersion> prodVersion = productVersionService.getProductVersion(prodVerId);   
+                prodVersion.ifPresent(x -> productVersions.add(x));
+            }
+        }
+        
+        if((productNames == null || productNames.isEmpty())  
+           && 
+           (productVersionIds == null || productVersionIds.isEmpty())){
+            productVersions.addAll(productVersionService.getAll());
+        }
+        
+        return productVersions;
+    }
+
+    private Map<WhiteArtifact, Set<ProductVersion>> getWhiteArtifacts(Set<ProductVersion> prodVers) {
+        Map<WhiteArtifact, Set<ProductVersion>> mapProdVersToArtifact = new HashMap<>();
+        for (ProductVersion currentProdVer : prodVers) {
+            for (WhiteArtifact wa : currentProdVer.getWhiteArtifacts()) {
+                Set<ProductVersion> prodVersContainingArtifact = mapProdVersToArtifact.get(wa);
+                if (prodVersContainingArtifact == null)
+                    prodVersContainingArtifact = new HashSet<>();
```

Comment:
You can use [Map#computeIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-) when getting existing or creating new entry in map.